File: Array/Python/07_Maximum_Subarray.py
# Maximum Subarray
# Question Link - https://leetcode.com/problems/maximum-subarray/


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
      
      # TC - O(n^2), SC - O(1)
      # maxi = nums[0]
      # for i in range(len(nums)):
      #     sum = 0
      #     for j in range(i, len(nums)):
      #         sum += nums[j]
      #         maxi = max(maxi, sum)

      
      # TC - O(n), SC - O(1)
        sum = 0
        maxi = nums[0]
        for i in range(len(nums)):
            sum += nums[i]
            maxi = max(maxi, sum)
            if sum < 0 :
                sum = 0

        return maxi

File: Array/Cpp/07_Maximum_Subarray.cpp
// Maximum Subarray
// Question Link - https://leetcode.com/problems/maximum-subarray/


class Solution {
public:
    int maxSubArray(vector<int>& nums) {

        // TC - O(n^2), SC - O(1)
        // int maxi = nums[0];
        // for(int i=0; i<nums.size(); i++){
        //     int sum = 0;
        //     for(int j=i; j<nums.size(); j++){
        //         sum += nums[j];
        //         maxi = max(maxi, sum);
        //     }
        // }
        // return maxi;

        // TC - O(n), SC - O(1)
        int sum = 0;
        int maxi = nums[0];

        for(int i=0; i<nums.size(); i++){
            sum += nums[i];
            maxi = max(maxi, sum);
            if(sum < 0){
                sum = 0;
            }
        }

        return maxi;
    }
};
